modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt

```diff
@@ -201,13 +185,9 @@ fun <R> Right(right: R): Either<Nothing, R> = Either.right(right)
  *
  * @param f The function to bind across [Either.Right].
  */
+@Suppress("UNCHECKED_CAST")
 inline fun <A, B, C> EitherOf<A, B>.flatMap(f: (B) -> Either<A, C>): Either<A, C> =
-  fix().let {
-    when (it) {
-      is Right -> f(it.b)
-      is Left -> it
-    }
-  }
+  fix().let { eith -> eith.fold({ eith as Left<A> }, f) }
```

Comment:
Is the cast necessary? If so, and given that it's pervasive, please add a `unsafeRecast` inline function that does this.

Reply:
This is not `unsafeRecast` this is an optimization to avoid re-wrapping `Left` values on short-circuits. Would you be in favor writing directly using tags here instead of an inlined fold then?

Reply:
I mean that there are several instances of this through the codebase, let's unify them in a single function.

And because you're returning an either from both sides, the cast may not be necessary in this specific instance as Left is defined with Nothing for the right side.

Reply:
The idea is that we can write a faster `fold` in terms of tags which is consistently faster to `when` because of the table switch optimization this fold becomes inline and the rest of the Arrow code base uses `fold` wherever is now using `when` with `is Left` etc.

Reply:
Yeah, that I understand. I mean literally a function `fun <A> Either<Nothing, Nothing>.recastLeftUnsafe(): Either<A, Nothing> = eith as Left<A>`